File: a2zSheet/step2_sorting/sorting.go
package main

import (
	"bufio"
	. "fmt"
	. "math"
	"os"
)

var (
	in  = bufio.NewReader(os.Stdin)
	out = bufio.NewWriter(os.Stdout)
)

func swap(a, b *int) {
	temp := *a
	*a = *b
	*b = temp
}

func SelectionSort(arr []int) {
	for i := 0; i < len(arr); i++ {
		mini := MaxInt32
		min_ind := -1
		for j := i; j < len(arr); j++ {
			if mini > arr[j] {
				mini = arr[j]
				min_ind = j
			}
		}
		swap(&arr[i], &arr[min_ind])
	}

	Print(arr)
}

func BubbleSort(arr []int) {
	for i := 0; i < len(arr); i++ {
		for j := 0; j < len(arr)-i-1; j++ {
			if arr[j] > arr[j+1] {
				swap(&arr[j], &arr[j+1])
			}

		}
	}
	Print(arr)
}

func InsertionSort(arr []int) {
	for i := 0; i < len(arr); i++ {
		for j := i; j > 0; j-- {
			if arr[j] < arr[j-1] {
				swap(&arr[j], &arr[j-1])
			}
		}
	}
	Print(arr)
}

func Merge(start, mid, end int, arr []int) {
	temp := make([]int, end-start+1)
	l, r := start, mid+1

	k := 0
	for l <= mid && r <= end {
		if arr[l] <= arr[r] {
			temp[k] = arr[l]
			l++
		} else {
			temp[k] = arr[r]
			r++
		}

		k++
	}
	for l <= mid {
		temp[k] = arr[l]
		l++
		k++
	}
	for r <= end {
		temp[k] = arr[r]
		r++
		k++
	}

	for i := start; i <= end; i++ {
		arr[i] = temp[i-start]
	}
}
func MergeSort(start, end int, arr []int) {
	if start >= end {
		return
	}
	mid := (start + end) / 2
	MergeSort(start, mid, arr)
	MergeSort(mid+1, end, arr)
	Merge(start, mid, end, arr)
}

func Partition(arr []int, start, end int) int {
	pivot := arr[start]
	l, r := start, end

	for l < r {
		for arr[l] <= pivot && l < end {
			l++
		}
		for arr[r] > pivot && r > start {
			r--
		}
		if l < r {
			swap(&arr[l], &arr[r])
		}
	}
	swap(&arr[start], &arr[r])

	return r
}

func QuickSort(arr []int, start, end int) {
	if start >= end {
		return
	}
	pivot := Partition(arr, start, end)
	QuickSort(arr, start, pivot-1)
	QuickSort(arr, pivot+1, end)
}

func solve() {
	var n int
	Fscan(in, &n)

	arr := make([]int, n)
	for i := 0; i < n; i++ {
		Fscan(in, &arr[i])
	}
	QuickSort(arr, 0, n-1)

	Print(arr)
}

func main() {
	var t int
	for Fscan(in, &t); t > 0; t-- {
		solve()
	}
}

File: a2zSheet/step2_sorting/sorting.cpp
#include<bits/stdc++.h>
using namespace std;

// Best Case O(N)
void BubbleSort(vector<int>&v){
    int n = v.size();
   
    for(int i = 0;i<n;i++){
        bool swapped = false;
        for(int j = 0;j<n-i-1;j++){
            if(v[j]>v[j+1]){
                swap(v[j],v[j+1]);
                swapped = true;
            }
        }
        if(!swapped){
            break;
        }
    }
}

void SelectionSort(vector<int>&v){
    int n = v.size();
    for(int i = 0;i<n;i++){
        int mini = v[i];
        int min_ind = i;
        for(int j = i+1;j<n;j++){
            if(mini>v[j]){
                mini = v[j];
                min_ind = j;
            }
        }
        swap(v[min_ind],v[i]);
    }
}

// Best Case O(N)
void InsertionSort(vector<int>&v){
    int n = v.size();
    for(int i = 0;i<n;i++){
        int j = i;
        while (j > 0 && v[j - 1] > v[j]) {
            swap(v[j-1],v[j]);
            j--;
        }
    }
}

void merger(int l,int mid,int h,vector<int>&v){
    int left = l;
    int right = mid+1;
    vector<int>temp;
    while(left<=mid and right<=h){
        if(v[left]<=v[right])temp.push_back(v[left++]);
        else{
            temp.push_back(v[right++]);
        }
    }
    while(left<=mid)temp.push_back(v[left++]);
    while(right<=h)temp.push_back(v[right++]);

    for(int i = l;i<=h;i++){
        v[i] = temp[i-l];
    }
}

// O(nlog(n)) in all the cases
void mergeSort(int l,int h,vector<int>&v){
    if(l>=h)return;
    int mid = (l+h)/2;
    mergeSort(0,mid,v);
    mergeSort(mid+1,h,v);
    merger(l,mid,h,v);
}

int Partition(int l,int h,vector<int>&v){
    int pivot = l;
    int i = l,j = h;
    while(i<j){
        while(i<=h-1 and v[i]<=v[pivot])i++;
        while(j>=l+1 and v[j]>=v[pivot])j--;
        if(i<j)swap(v[i],v[j]);
    }
    swap(v[j],v[pivot]);
    return j;
}

/*
Variation	Time Complexity	Space Complexity
Best Case	  O(n log n)	O(log n)
Average Case  O(n log n)	O(log n)
Worst Case	  O(n^2)	    O(n)  Reverse Array is provided.

*/

void QuickSort(int l,int h,vector<int>&v){
    if(l>=h)return;
    int pivot = Partition(l,h,v);
    QuickSort(l,pivot-1,v);
    QuickSort(pivot+1,h,v);
}


int main(){
    int n;cin>>n;
    vector<int>v(n);
    for(auto &c:v)cin>>c;
    SelectionSort(v);
    for(auto &c:v)cout<<c<<" ";
    cout<<'\n';
    return 0;
}